### PR TITLE
add more scheduling unit tests

### PR DESCRIPTION
**Pull request by tzneal**
**1. Issue, if available:**

N/A

**2. Description of changes:**

Adds additional unit tests to scheduling:
- ensures that all instance types passed to the cloud provider meet
  the required constraints
- verify prov/pod constraints on OS
- verify prov/pod constraints on zone/capacity type
- verify mixed prov/pod constraints

**3. How was this change tested?**

Unit test, no functional changes were made.

**4. Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: *link to issue*
- [X] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/controllers/provisioning/scheduling/suite_test.go [hunk 1]
```diff
@@ -2033,14 +2034,24 @@ var _ = Describe("Binpacking", func() {
...
 })
-
 var _ = Describe("Instance Type Selection", func() {
 	var minPrice float64
 	var instanceTypePrices map[string]float64
 	nodePrice := func(n *v1.Node) float64 {
 		return instanceTypePrices[n.Labels[v1.LabelInstanceTypeStable]]
 	}
 
+	asString := func(s sets.String) string {
```
Comment (ellistarn) [inline pkg/controllers/provisioning/scheduling/suite_test.go:2044]: Can you use `strings.Join()`?

File: pkg/controllers/provisioning/scheduling/suite_test.go [hunk 2]
```diff
@@ -2053,25 +2064,37 @@ var _ = Describe("Instance Type Selection", func() {
...
-						OperatingSystems: sets.NewString("linux", "windows", "darwin"),
-						Resources: v1.ResourceList{
-							v1.ResourceCPU:    resource.MustParse(fmt.Sprintf("%d", cpu)),
-							v1.ResourceMemory: resource.MustParse(fmt.Sprintf("%dGi", mem)),
-						},
-					})
-					instanceTypePrices[it.Name()] = it.Price()
-					if it.Price() < minPrice {
-						minPrice = it.Price()
+			for _, mem := range []int{1, 2, 4, 8, 16, 32, 64, 128} {
```
Comment (ellistarn) [inline pkg/controllers/provisioning/scheduling/suite_test.go:2070]: These suites are getting pretty big. Thoughts on separating them out into multiple logical suites?

Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/suite_test.go:2070]: Ginkgo only lets you have one suite per package as far as I can tell.  I can at least separate it into multiple files though.

File: pkg/controllers/provisioning/scheduling/suite_test.go [hunk 3]
```diff
@@ -2053,25 +2064,37 @@ var _ = Describe("Instance Type Selection", func() {
...
+				for _, zone := range []string{"test-zone-1", "test-zone-2"} {
```
[the remark this answers is not in the document]
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/suite_test.go:2071]: Yeah, that could be useful other places.

**Comment by tzneal** [comment]
> I like this! Do you think there would be value in making a script that loads the real AWS instance types in the AWS cloud provider?

Yeah, that would be great! I don't think ginkgo supports table driven tests, but we could make a test helper that lets you pass in the list of instance types. One could be this unique set and the other are a snapshot of what real instance types were.  That would be good to hook into the AWS CP EC2 Instance Type -> cloudprovider.InstanceType translation as well to get more testing on that.

File: pkg/cloudprovider/fake/instancetype.go [hunk 4]
```diff
@@ -65,6 +68,40 @@ func NewInstanceType(options InstanceTypeOptions) *InstanceType {
...
 }
 
+// InstanceTypesAssorted create 1000+ unique instance types with varying CPU/memory/architecture/OS/zone/capacity type.
+func InstanceTypesAssorted() []cloudprovider.InstanceType {
+	var instanceTypes []cloudprovider.InstanceType
+	for _, cpu := range []int{1, 2, 4, 8, 16, 32, 64} {
+		for _, mem := range []int{1, 2, 4, 8, 16, 32, 64, 128} {
+			for _, zone := range []string{"test-zone-1", "test-zone-2"} {
+				for _, ct := range []string{v1alpha1.CapacityTypeSpot, v1alpha1.CapacityTypeOnDemand} {
+					for _, os := range []sets.String{sets.NewString("darwin"), sets.NewString("linux"), sets.NewString("windows")} {
```
Comment (bwagner5) [inline pkg/cloudprovider/fake/instancetype.go:78]: can we remove `darwin`?

File: pkg/controllers/provisioning/scheduling/instance_selection_test.go [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+			test.PodOptions{NodeRequirements: []v1.NodeSelectorRequirement{{
+				Key:      v1.LabelTopologyZone,
+				Operator: v1.NodeSelectorOpIn,
+				Values:   []string{"test-zone-2"},
+			}}}))
+		node := ExpectScheduled(ctx, env.Client, pod[0])
+		Expect(nodePrice(node)).To(Equal(minPrice))
+		ExpectInstancesWithLabel(cloudProv.CreateCalls[0].InstanceTypes, v1.LabelTopologyZone, "test-zone-2")
+	})
+	It("should schedule on one of the cheapest instances (prov ct = spot)", func() {
```
Comment (felix-zhe-huang) [inline pkg/controllers/provisioning/scheduling/instance_selection_test.go:207]: Just to follow the above format, should we also test the case (pod ct = spot)?

File: pkg/cloudprovider/fake/instancetype.go [hunk 6]
```diff
@@ -65,6 +68,40 @@ func NewInstanceType(options InstanceTypeOptions) *InstanceType {
 	}
 }
 
+// InstanceTypesAssorted create 1000+ unique instance types with varying CPU/memory/architecture/OS/zone/capacity type.
+func InstanceTypesAssorted() []cloudprovider.InstanceType {
+	var instanceTypes []cloudprovider.InstanceType
+	for _, cpu := range []int{1, 2, 4, 8, 16, 32, 64} {
+		for _, mem := range []int{1, 2, 4, 8, 16, 32, 64, 128} {
+			for _, zone := range []string{"test-zone-1", "test-zone-2"} {
```
Comment (bwagner5) [inline pkg/cloudprovider/fake/instancetype.go:76]: would adding a 3rd zone break anything? 3 zones is a more typical "real world" setup

**Review by bwagner5 — COMMENTED** [review]
just a couple questions, nothing mandatory

File: pkg/controllers/provisioning/scheduling/instance_selection_test.go (hunk 5 above)
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/instance_selection_test.go:207]: We are but I labeled the test wrong 😆 fixed.

File: pkg/cloudprovider/fake/instancetype.go (hunk 4 above)
Comment (tzneal) [reply pkg/cloudprovider/fake/instancetype.go:78]: SGTM

File: pkg/cloudprovider/fake/instancetype.go (hunk 6 above)
Comment (tzneal) [reply pkg/cloudprovider/fake/instancetype.go:76]: If it does, our code is broken, adding a third zone.

File: pkg/controllers/provisioning/scheduling/instance_selection_test.go [hunk 7]
```diff
@@ -0,0 +1,453 @@
...
+				for _, it := range cloudProv.CreateCalls[0].InstanceTypes {
+					totalReserved := resources.Merge(totalPodResources, it.Overhead())
+					// the total pod resources in CPU and memory + instance overhead should always be less than the
+					// resources available on every viable instance has
+					Expect(totalReserved.Cpu().Cmp(it.Resources()[v1.ResourceCPU])).To(Equal(-1))
+					Expect(totalReserved.Memory().Cmp(it.Resources()[v1.ResourceMemory])).To(Equal(-1))
+				}
+			}
+		}
+	})
```
Comment (felix-zhe-huang) [inline pkg/controllers/provisioning/scheduling/instance_selection_test.go:405]: If I understand correctly, the above cases can all find compatible instance type for the give pod. Can we add some cases with multiple pods that will create incompatibility and we should see them been put on separated nodes?

Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/instance_selection_test.go:405]: This set of tests was geared at two recent issues that caused point releases where we were passing a set of instance types to the cloudprovider where some, but not all of the instance types matched the constraints. We have some individual tests for that behavior:
- "should launch pods with resources that aren't on any single instance type on different instances"
- "should schedule incompatible pods to the different node"

File: pkg/cloudprovider/fake/instancetype.go [hunk 8]
```diff
@@ -65,6 +68,40 @@ func NewInstanceType(options InstanceTypeOptions) *InstanceType {
 	}
 }
 
+// InstanceTypesAssorted create many unique instance types with varying CPU/memory/architecture/OS/zone/capacity type.
+func InstanceTypesAssorted() []cloudprovider.InstanceType {
+	var instanceTypes []cloudprovider.InstanceType
+	for _, cpu := range []int{1, 2, 4, 8, 16, 32, 64} {
```
[the remark this answers is not in the document]
Comment (tzneal) [reply pkg/cloudprovider/fake/instancetype.go:74]: Good idea, I added a few more tests.

**Review by felix-zhe-huang — APPROVED** [review]
LGTM!

**Pull request merged**